enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,269 @@
+
+Mitigation: The data gathering requests will be queued and only one gathering at time will be allowed. 
+The original requirement for the external processing was to provide the most recent Insights analysis and recommendations in the HAC (Hybrid Application Console) and ACM (Advanced Cluster Manager) and there is no use case for maintaining a history of Insights recommmedations/analyses in the cluster, plus it would take considerable effort to enable such a history. 
+
+### Drawbacks
+
+- Very frequent data gathering in a cluster can produce a significant load on the processing part in the console.redhat.com and can overwhelm the processing pipeline. 
+Theoretically, we could limit the number of user-trigerred gathering over time, but this problem can happen today when something is deleting the Insights operator pod on a regular basis with appropriate timing (so that data can be uploaded).  
+- When the newly spawned container is failing to upload the data then it can be running for some longer time (couple of hours). 
+
+
+## Design Details
+
+### Open Questions
+- ~~How to solve the problem with only one Insights analysis result available in a cluster? How to avoid the "uncertainty" of the Insights analysis results described in the [Risk and Mitigations](#risks-and-mitigations) section?~~ 
+  It would make sense to queue the data gathering requests.
+- ~~Should the newly created gathering process update the clusteroperator conditions (e.g it fails to upload the data, should it update the `Degraded` status as it is today)? This can probably be racy too or it can lead to temporary and inconsistent conditions state.~~ 
+  Yes. The scenario is described in the [Updates required to allow on-demand data gathering](#updates-required-to-allow-on-demand-data-gathering) section.
+- ~~What all is required to spawn an extra container? Do we need a new image and a new binary (note that the Insights operator currently includes some additional features - such as pulling SCA certificates - that are not required for a new container/process)?~~ I think we can use the existing image and binary and introduce a new command.
+- ~~Is some clean-up of the old `datagather.insights.openshift.io` CRs (and completed containers/pods) required? I think it would make sense to keep them only for 24h.~~ The CRs older than 24 hours will be removed by the Insights operator. This time option will be configurable via the operator configuration. 
+
+### Test Plan
+
+This will require significant updates to the existing Insights operator testsuite. 
+
+### Graduation Criteria
+
+The plan is to introduce the new API in version `v1alpha1`. This will probably require to remove the `TechPreviewNoUpgrade` limitation for the existing `insightsdatagather.config.openshift.io` (see the [config API enhancement](./insights-config-api.md)).
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+TBD 
```

Comment:
It seems it would be simpler to introduces this API as `v1alpha1` without `TechPreviewNoUpgrade` (see the `Graduation criteria`) so I am not sure what to do with these two sections.

Reply:
OK it can't be `v1alpha1` without `TechPreviewNoUpgrade` so I would suggest `v1beta1`. My concern is that refactoring the existing Insights operator to create jobs is not so small change and having both versions (the original and the new) with the "techpreview switch" would be somewhat difficult.

Reply:
I was experimenting with the refactoring last week and it should be feasible to have the common behavior by default and allow the new approach with the "techpreview switch". So TP it is.